### PR TITLE
i18n tooling + translation updates

### PR DESCRIPTION
**Pull request by OmarIthawi**
We'd like this XBlock to be translated to Japanese and other languages, I've also supported JavaScript translations via 
[`django-statici18n`](https://github.com/zyegfryed/django-statici18n/pull/41).

Before you continue, take a look at [edx/i18n-tools](https://github.com/edx/i18n-tools) and [edx/cookiecutter-django-app](https://github.com/edx/cookiecutter-django-app).

### Summary of the Changes
 - Added a Makefile to help manage the translations. The file is copied and modified from: [edx/cookiecutter-django-app](https://github.com/edx/cookiecutter-django-app)
 - Created a `locale` link to `oppia/translations` to allow the `i18n_tools` find the po/mo files
 - Added `oppia/translations/config.yaml` which works as the configuration file of the `i18n_tools` command 
 - Added `requirements-dev.txt` to pin development packages. Used for now only for translations.

### TODO

 - [x] Support JavaScript translations as `textjs.po`
 - [x] Document how to update the strings properly from Transifex
 - [x] Manually test that the the dummy Esperanto translation is working

#### Studio
image

#### LMS
image

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,60 @@
+.PHONY: dummy_translations extract_translations fake_translations help pull_translations push_translations \
```
Comment (seanlip) [inline Makefile]: Try to avoid backslashes at ends of lines.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+quality: ## check coding style with pycodestyle and pylint
+	echo 'TODO: Refactor quality checks from .travis.yml into this Makefile'
+	exit 1
+
+requirements: ## install development environment requirements
+	pip install -r requirements.txt --exists-action w
+	pip install -r requirements-dev.txt --exists-action w
+
+test: clean ## run tests in the current virtualenv
```
Comment (seanlip) [inline Makefile:35]: Use a similar style throughout the makefile: name, colon, comment, then the commands start on the next line. Ditto below.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+detect_changed_source_translations:
+	i18n_tool changed
+
+pull_translations: dummy_translations ## pull translations from POEditor.com
+	poeditor pull
+	make compile_translations
+	make clean
+
+push_translations: clean extract_translations ## push source translation files (.po) from POEditor.com
```
Comment (seanlip) [inline Makefile]: from --> to

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	exit 1
+
+requirements: ## install development environment requirements
+	pip install -r requirements.txt --exists-action w
+	pip install -r requirements-dev.txt --exists-action w
+
+test: clean ## run tests in the current virtualenv
+	python run_tests.py
+
+selfcheck: ## check that the Makefile is well-formed
```
Comment (seanlip) [inline Makefile:38]: This doesn't seem to do any actual checking?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	@echo "The Makefile is well-formed."
+
+## Localization targets
+
+extract_translations: ## extract strings to be translated, outputting .po files
+	rm -f locale/en/LC_MESSAGES/django.po
+	django-admin.py makemessages -l en -d django
```
Comment (seanlip) [inline Makefile]: Not sure where django-admin and i18n_tool are coming from?

**Review by seanlip — CHANGES_REQUESTED** [review]
Hi @OmarIthawi -- firstly, thank you for making a contribution to the Oppia repository!

Unfortunately, it's been a while since I've looked at the code for this, and I am a little uncertain about a bunch of the edits that you did. (I only left comments in the Makefile but on glancing through the other files, I noticed that there were things I had questions about there too.) However, I don't have the bandwidth at the moment to set up openedx again and test it out thoroughly, or to do a full in-depth review of this. At the same time, though, I agree with you that this might be useful for other devs in the future.

So, here's what I propose: would you be up for creating your own fork of this repository, and augmenting it with translations functionality? Then, we can add a link to your repo in the README, so that other people know about it. Please let us know if this sounds good.

Thanks!

**Comment by OmarIthawi** [comment]
> Hi @OmarIthawi -- firstly, thank you for making a contribution to the Oppia repository!

@seanlip Thanks for the review. I'll address/answer your inline comments shortly.

> Unfortunately, it's been a while since I've looked at the code for this, and I am a little uncertain about a bunch of the edits that you did. However, I don't have the bandwidth at the moment to set up openedx again and test it out thoroughly, or to do a full in-depth review of this. 

I totally understand. So, at Appsembler we (sort of) own maintenance for this XBlock, as long us we're offering it as part of our [Open edX SaaS](https://www.appsembler.com/tahoe/). That's the case for now.

Therefore, we can provide some review capacity from one of our engineers as well.

> At the same time, though, I agree with you that this might be useful for other devs in the future.
>
> So, here's what I propose: would you be up for creating your own fork of this repository, and augmenting it with translations functionality? Then, we can add a link to your repo in the README, so that other people know about it. Please let us know if this sounds good.

I've already [forked the repo](https://github.com/appsembler/oppia-xblock), in order to make this pull request. However, I'm following an upstream first approach, in which I'd try to make less repositories and keeping all community contributions (in this case Oppia and Appsembler) in one place (this repo preferably).

Unless, you'd like to deprecate this repository in favor of our own: https://github.com/appsembler/oppia-xblock

Would it be possible to provide me with write access for me to this repo? If so, then it won't matter much if you don't have time, as we can do the review as well.

What do you think?

File: Makefile (hunk 4 above)
Comment (OmarIthawi) [reply Makefile:38]: This is step to do exactly nothing, but fails if the Makefile isn't well formed.

File: Makefile (hunk 3 above)
Comment (OmarIthawi) [reply Makefile]: Done

File: Makefile (hunk 5 above)
Comment (OmarIthawi) [reply Makefile]: It's installed via `$ make requirements` which installs `requirements-dev.txt`. Those requirements are a standard requirement for Open edX related applications.

File: Makefile (hunk 1 above)
Comment (OmarIthawi) [reply Makefile]: It was a typo. Resolved.

File: Makefile (hunk 2 above)
Comment (OmarIthawi) [reply Makefile:35]: I'm using the [edX's cookiecutter](https://github.com/edx/cookiecutter-django-app/blob/master/%7B%7Bcookiecutter.repo_name%7D%7D/Makefile) and that's a convention there, which I'd like to stay very close to. The goal is to make this XBlock as convenient to reason about as possible for other Open edX developers.

**Comment by seanlip** [comment]
Hi @OmarIthawi, I thought about this, and would suggest the following:

- I think we should maintain ownership of the original XBlock, since we own the repository that it embeds. In general, we would be happy to accept PRs for bugs in the XBlock, but at this time we are not taking feature requests. The XBlock we wrote is meant to be a simple embedder.
- In general, I think that code submitted to any repository should be reviewed and accepted by an owner of that repository. The current change makes the existing XBlock much more complex, and from taking a quick look at it, I'm not comfortable accepting it as-is. I think it will require at least some rounds of review, but unfortunately, we don't have the bandwidth to review it at the moment. 
- On the other hand, I don't want to block your workflow. So, I think that the best interim solution here is for Appsembler to maintain its own version of the XBlock, and, if you like, we'd be happy to add a link in the README to https://github.com/appsembler/oppia-xblock. That way, you can maintain and update it without us blocking things, and if your work is useful to others in the community, they will be able to find it.
- At some point in the future, we might be able to accept responsibility for i18n, and at that point we may be able to do a merge of what you've done with the current xblock repo. But I don't think that time has come yet.

I hope that works? Again, thank you for starting off by submitting it here, and I'm sorry that we won't be able to accept it at this time.

**Comment by OmarIthawi** [comment]
Hi @seanlip! Thanks for taking your time to comment. What you're suggesting makes sense, and works well in the short term. 

I'll close this PR for now in favor of the fork pull request (https://github.com/appsembler/oppia-xblock/pull/1). I appreciate adding a link in the README.md.

Meanwhile, I'd like to explain why I'm doing this. As an open-source contributor, I aim to make code-changes upstream-first, so forking becomes the exception. It's something that I've learned from [RedHat's upstream-first approach](https://opensource.com/article/16/12/why-red-hat-takes-upstream-first-approach) and it's one of the success factors in their open-source based business model.

**Comment by seanlip** [comment]
Hi @OmarIthawi, sounds good, thanks!

I agree with upstream-first, too, in general, and I definitely appreciate your sending the code here first. I'm sorry about our lack of bandwidth to support this more fully at this point, and wanted to say thanks for understanding.

Please let me know when you've got the other PR in, and we'd be happy to add the link on our end as discussed.

Thanks again!

**Comment by seanlip** [comment]
Updated the README. Thanks!

**Pull request closed without merging**